### ISSUE TITLE
Stutter on iOS devices caused by touch input events

### ISSUE THREAD

**Issue by AlexRixhardson** [godotengine/godot#32139]
**Godot version:**
3.1.1. stable

**OS/device including version:**
iOS 12 - iPad Mini 2

**Issue description:**
Game works with solid 60 FPS until player starts touching the screen. FPS drops to approx. 57 once the player starts touching the screen which causes apparent stuttering. Cannot reproduce on an Android device.

**Steps to reproduce:**
- Run minimal reproduction project on an iPad Mini 2 or an equivalent device (do not run on high end devices, such as iPhone X, as this problem cannot be reproduced on them)
- Start dragging fingers across the screen and after a few seconds you should start seeing FPS drop

**Minimal reproduction project:**
[Archive.zip](https://github.com/godotengine/godot/files/3613033/Archive.zip)

**Comment by AlexRixhardson** [comment]
Some additional info: the problem is present in 3.2 master branch as well.

I tried to comment out some of the code from Godot (hoping that stutter would disappear), but to no avail. The stutter is still present in low-end device (such as iPad Mini 2) when player drags finger across the screen, even though dragging is effectively not working anymore after commenting out all of the following:

in **/platform/iphone/gl_view.mm**:
```
		// Process all input events
		// while (CFRunLoopRunInMode(kCFRunLoopDefaultMode, 0, TRUE) == kCFRunLoopRunHandledSource);
```
also in **/platform/iphone/gl_view.mm**:
- complete method `(void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event`

in **/platform/iphone/os_iphone.cpp**:
- complete body of method `OSIPhone::touch_drag`
- line `input->parse_input_event(event_queue[i]);` in function `OSIPhone::iterate()`

It seems that I am looking at the wrong pieces of code for the cause of the stuttering issue. Does anyone who is more familiar with the structure of Godot's iOS code know where roughly could the cause be? Where could I look? :-/

**Comment by KoBeWi** [comment]
Can anyone still reproduce this bug in [Godot 3.2.3 rc4](https://godotengine.org/article/release-candidate-godot-3-2-3-rc-4) or any later release?

**Comment by akien-mga** [comment]
No answer, so closing. Please comment if you can still reproduce it, and make sure that an up-to-date reproduction project is included.

**Comment by djrain** [comment]
Seems this issue or something similar is still present in 3.5.

 Our game uses tap controls, and we're experiencing noticeable stuttering (drops to 56-58 fps) while tapping on iPhone 6s. Whereas skipping the input and triggering the same player movement through code runs perfectly smooth.

Also, Xcode shows a memory leak. In a scene containing only an FPS label, the app's memory use increases precisely when tapping the screen and doesn't go back down. It only increases on taps (not hold/swipe), and increases faster as more fingers are used. FPS degrades more and more as the tapping continues, but stays solid at 60 when no input is happening.

<img width="850" alt="Screen Shot 2022-09-10 at 3 19 58 PM" src="https://user-images.githubusercontent.com/33777501/189504166-911cc123-091e-4a20-813f-a2dd56082663.png">
The memory started at 131 MB and increased to 144 within a few minutes of tapping.



It's really a default project apart from the FPS label, but here it is:
[iOSTouchStutter.zip](https://github.com/godotengine/godot/files/9541482/iOSTouchStutter.zip)

**Comment by Calinou** [comment]
cc @bruvzg

**Comment by djrain** [comment]
Not to rush things, but would anyone perhaps have a chance to look into this soon? Unfortunately this issue is kinda preventing us from releasing our game as the stutters are quite bad, and we haven't found any workaround. I went to try and test this in Godot 4 beta 1 to see if it's present there, but the iOS builds are currently broken so I couldn't tell.

I'm happy to poke around myself, but I'm not familiar with the internals here - any hints or direction would be appreciated!

**Comment by tbveralrud** [comment]
I believe #69200 fixed this, but I'll let @djrain verify this claim.

**Comment by djrain** [comment]
Yes, our testing suggests that #69200 fixes this.

**Comment by djrain** [comment]
Unfortunately, it seems I was mistaken, or my previous test was a fluke somehow - there is still a serious stuttering problem.

In 3.6 beta 1, the memory leak is fixed, but tapping causes FPS to drop significantly (using same MRP I uploaded previously). I also re-tested in 3.5.1 with #69200 and the stuttering is present there too. This is on an iPhone 14 Pro Max.

Is anyone able to confirm this?

**Comment by tbveralrud** [comment]
I'll look soon. Will do so on an iPhone 12 Pro.

**Comment by tbveralrud** [comment]
I am not able to reproduce so far. (iPhone 12 Pro / iOS 16.3.1)
Tested with a very recent 3.x commit (17e0a02).

When idle, the label will consistently read 60 or 59 fps. These values are consistent even when rapidly tapping. I added a tap counter and have tested beyond 4000 taps. Do I need to go far beyond this?

![tap counter](https://user-images.githubusercontent.com/1694668/233899559-713b48b7-3783-407f-b82f-35fffc9c483f.jpeg)

Possibly related: Xcode gives me a runtime warning that the user interactive thread is blocked by a lower priority thread.
<img width="493" alt="hang risk" src="https://user-images.githubusercontent.com/1694668/233899094-df2e621b-452e-4828-8eaa-c39204cb0374.png">

I could investigate this lead to see if it addresses the stutter but it would be a blind fix without a repro.

**Comment by djrain** [comment]
> I added a tap counter and have tested beyond 4000 taps. Do I need to go far beyond this?

Thanks for testing! No, there's no need to go further. For me, it's quite easy to get FPS to drop significantly (sometimes even below 50) and it happens right from the start. 

Here is some quick Xcode profiling I recorded. You can see it starts out normal, with no input. When I start tapping rapidly with three fingers, FPS drops and both CPU and GPU usage increase. Then everything returns to normal after I stop tapping.

https://user-images.githubusercontent.com/33777501/233914675-7b1e29ab-e110-4095-8163-e4c03d36ac1a.mov


> Possibly related: Xcode gives me a runtime warning that the user interactive thread is blocked by a lower priority thread.

I've been seeing that as well, bit concerning but personally know nothing about it :)

**Comment by djrain** [comment]
I suspect the stutter I'm experiencing is specific to devices with Pro Motion display. That would explain why we're not seeing it on iPhone 12 or my iPhone 6s. I've made a new issue #76425 since I do think this particular issue was actually solved.

**Comment by tbveralrud** [comment]
That's a very good hunch. I'll keep thinking on it.

FWIW, I tested a couple more devices / iOS versions today and couldn't reproduce.
iPad (9th gen) / 16.4.1
iPhone 7 / 15.2.1
